doc/man/lttng-ust.3.txt

```diff
@@ -0,0 +1,1030 @@
+
+
+[[tracepoint-loglevel]]
+`TRACEPOINT_LOGLEVEL()` usage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Optionally, a *log level* can be assigned to a defined tracepoint.
+Assigning different levels of severity to tracepoints can be useful:
+when controlling tracing sessions, you can choose to only enable
+events falling into a specific log level range using the
+nloption:--loglevel and nloption:--loglevel-only options of the
+man:lttng-enable-event(1) command.
+
+Log levels are assigned to tracepoints that are already defined using
+the `TRACEPOINT_LOGLEVEL()` macro. The latter must be used after having
+used `TRACEPOINT_EVENT()` or `TRACEPOINT_EVENT_INSTANCE()` for a given
+tracepoint. The `TRACEPOINT_LOGLEVEL()` macro is used as follows:
+
+------------------------------------------------------------------------
+TRACEPOINT_LOGLEVEL(
+    /* tracepoint provider name */
+    my_provider,
+
+    /* tracepoint/event name */
+    my_tracepoint,
+
+    /* log level */
+    TRACE_INFO
+)
+------------------------------------------------------------------------
+
+The available log level definitions are:
+
+include::log-levels.txt[]
+
+See the <<example,EXAMPLE>> section below for a complete example.
+
+
+[[tracepoint]]
+Instrumenting your application
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Once the tracepoint provider is created (see the
+<<creating-tp,Creating a tracepoint provider>> section above), you can
+instrument your application with the defined tracepoints thanks to the
+`tracepoint()` macro:
+
+[verse]
+#define *tracepoint*('prov_name', 't_name', ...)
+
+With:
+
+'prov_name'::
+    Tracepoint provider name.
+
+'t_name'::
+    Tracepoint/event name.
+
+`...`::
+    Tracepoint arguments, if any.
+
+Make sure to include the tracepoint provider header file anywhere you
+use `tracepoint()` for this provider.
+
+NOTE: Even though LTTng-UST supports `tracepoint()` call site duplicates
+having the same provider and tracepoint names, it is recommended to use
+a provider/tracepoint name pair only once within the application source
+code to help map events back to their call sites when analyzing the
+trace.
+
+Sometimes, arguments to the tracepoint are expensive to compute (take
+call stack, for example). To avoid the computation when the tracepoint
+is disabled, you can use the `tracepoint_enabled()` and
+`do_tracepoint()` macros:
+
+[verse]
+#define *tracepoint_enabled*('prov_name', 't_name')
+#define *do_tracepoint*('prov_name', 't_name', ...)
+
+`tracepoint_enabled()` returns a non-zero value if the tracepoint
+named 't_name' from the provider named 'prov_name' is enabled at
+run time.
+
+`do_tracepoint()` is like `tracepoint()`, except that it doesn't check
+if the tracepoint is enabled. Using `tracepoint()` in such a scenario is
+dangerous since it also contains the `tracepoint_enabled()` check, thus
+a race condition is possible if the tracepoint is enabled between the
+user check and the `tracepoint()` macro invocation.
```

Comment:
yes, but we really don't care about this. That would be optimizing for a very narrow race window, which is useless.

The reason why we typically don't want to see tracepoint_enabled() used with tracepoint() is because we would perform the "enabled" check twice on the fast-path.